### DISCUSSION TITLE
Benchmarks, Tests, and Observations on AMP BC-250 APU

### DISCUSSION THREAD

**Discussion by mxreyer** [artyom-beilis/dlprimitives#44] [Show and tell]
Hi Artyom, thank you for maintaining this amazing project!

I have been playing around with my [AMD BC-250](https://elektricm.github.io/amd-bc250-docs/), a crypto-mining board running a repurposed PS5 chip with a powerful APU (after [unlocking](https://github.com/duggasco/bc250-40cu-unlock)and [overclocking](https://elektricm.github.io/amd-bc250-docs/bios/overclocking/), my machine has 40 CUs running at 2 GHz, 16GB GDDR6 memory with ~450 GB/s theoretical bandwidth - it should roughly be comparable with something like a NVIDIA T4 for compute workloads). This hardware has become quite popular in the Steam gaming community. I wanted to see whether I can use this box as a PyTorch ML server. Unfortunately, while Vulkan runs well on this box, the ROCm stack does not - but OpenCL through Mesa's rusticl runs fine. I actually got PyTorch running through your other project [artyom-beilis/pytorch_dlprim](https://github.com/artyom-beilis/pytorch_dlprim). It is amazing to see how your efforts allow us to run workloads on quirky hardware like this, massive kudos for that! Here, I wanted to contribute my performance benchmarks, plus some observations. 

**First the benchmarks.** I needed to set `#define CUSTOM_REDUCE 1` in `src/kernels/reduce.h` to get this to run (rusticl doesn't implement the optional OpenCL 3.0 work-group collectives, so `work_group_reduce_add` fails with `use of undeclared identifier`). The results can be found [in this gist](https://gist.github.com/mxreyer/e09c74a83a790a0c0aab57a9dfbf9ecd#file-benchmark-baseline-txt).

**Second, my observations.** When I compared ResNet-9 (CIFAR-10) performance in PyTorch (through [artyom-beilis/pytorch_dlprim](https://github.com/artyom-beilis/pytorch_dlprim)) to a T4 on Google Colab, I saw slightly disappointing results, especially in training (919 img/s vs 2,294 img/s). As I am not very proficient with GPU profiling myself, I asked Claude to find and fix bottlenecks in the current implementation, just for fun. I arrived at 2,337 img/s. The fixes target the following aspects of Winograd:
  - occupancy: improve the backward-filter split-K heuristic, which left 24 of 40 CUs idle on typical layers.
  - atomic add: RDNA1/2 have no hardware fp32 atomic add, so the backward kernels' accumulation was rewritten to not need atomics.
  - coalescing: filter layout, backward-filter tile mapping and the BatchNorm sums all had neighbouring lanes reading far-apart addresses.
  - prefetch: fetch the next tile while multiplying the current one.
  - LDS padding: dropped in the backward kernels, where it cost a second resident work-group per compute unit.
Disclaimer: Unfortunately, I cannot cross-check these performance gains against another APU/GPU.

Re-running the benchmarks from this repo with the fixes applied, I get the results from [this gist](https://gist.github.com/mxreyer/e09c74a83a790a0c0aab57a9dfbf9ecd#file-benchmark-patched-txt) - In training, vgg16 improves most, resnet50 and resnet18 also improve a lot. In inference, vgg16 and resnet18 benefit significantly.

**Third, on the unit tests.** I need to add the line `barrier(CLK_LOCAL_MEM_FENCE);` after `myval = my_reduce[0]; \` in `src/kernels/reduce.h` to get the `softmax` and `log_softmax` to pass (work-item 0 can start the next reduction and overwrite `my_reduce[0]` before slower work-items have read it; softmax reduces twice back to back). The test `conv2d_win` still fails, but it looks like a single output pixel sitting near zero and landing on the other side than the CPU reference does: the forward output passes, but ReLU masks that pixel's gradient differently, and the input gradient differs on exactly the 3×3 patch it feeds. See the test output [in this gist](https://gist.github.com/mxreyer/e09c74a83a790a0c0aab57a9dfbf9ecd#file-unittests-fixed-softmax-txt) (this run already had the `barrier(CLK_LOCAL_MEM_FENCE)` fix but none of the performance patches applied).

Would you be interested in me sharing more details about the performance improvements and the code fixes? I apologize that this is Claude, but GPU profiling is a little over my head. I did verify and measure all fixes on my own hardware. My repo github.com/mxreyer/pytorch-dlprim-gfx1013 has the patch series and a ready-to-file report for each fix; happy to open them as issues or PRs here if useful.

Cheers, and thanks again for providing us with this useful project!

**Comment by artyom-beilis** [comment]
First of all very interesting

> atomic add: RDNA1/2 have no hardware fp32 atomic add, so the backward kernels' accumulation was rewritten to not need atomics.


Can you elaborate on this?

I honestly had hart time implementing it without atomic because you need summing coltoim and for wingorad it is even harder
So if there an efficient implementation that works better without atomic  it would be very interesting



> Would you be interested in me sharing more details about the performance improvements and the code fixes? I apologize that this is Claude,

As long as you/me understand each and every  line and reason I have no problem whatsoever

There is huge number of relatively simple operators that just need to be implemented in PyTorch and  dlprimitives/onnx that probably can be done quite well by these tools.

And the project really needs them


> occupancy: improve the backward-filter split-K heuristic, which left 24 of 40 CUs idle on typical layers.


Interesting anyway there is built in  profiling in pytorch/dlprimitives  that can give you time per kernel so you can see what happens

I hadn't managed to get dedicated profiles yet but what you are saying is very interesting 

Looking forward!

**Reply by mxreyer** [reply]
Hi Artyom, thanks for the response!

> Can you elaborate on this?

The fix does not touch the GEMM path (col2im), only Winograd, and only applies for devices without a hardware fp32 atomic add where atomic add is emulated (specifically, when the device is not NVIDIA and doesn't advertise `cl_ext_float_atomics`). In backward-filter and backward-data, the patch identifies subsets of operations that do not touch each other, makes them write into their own data copy ("plane") in the operators's workspace, and introduces kernels `winconv_3x3_bwd_filter_reduce` and `winconv_3x3_bwd_data_reduce` that sum these copies.

For backward-filter, the planes are the split-K slices. For backward-data they are four parity classes of tiles that do not touch each other.

**Backward-filter:**

Within one slice z of the split-K, each of the dw elements is computed by exactly one work-item, so plain stores are safe. Slice z writes its 3x3 partial into plane z of the workspace. For each of the 9·C·N elements, `winconv_3x3_bwd_filter_reduce` then sums that element across s planes and applies beta (which also makes the separate beta pre-scale pass unnecessary). The workspace is s·9·C·N floats, and s is halved until that stays under 64 MB.

**Backward-data:**

The writers that collide are neighbouring tiles, whose 4 × 4 footprints overlap. The patch groups tiles so that no two tiles in a group overlap.

**In 1D**, tile t covers rows 2t−1, ... 2t+2. This means that all tiles of a given parity (t even, t odd) are pairwise disjoint:

```
 rows:      -1  0  1  2  3  4  5  6  7  8  9 10 11 12
 even  t=0:  #  #  #  #
 even  t=2:              #  #  #  #
 even  t=4:                          #  #  #  #
 odd   t=1:        #  #  #  #
 odd   t=3:                    #  #  #  #
 odd   t=5:                                #  #  #  #
```

Unless we are at the edges, each row is covered by exactly 2 tiles,

$$
t_1 = \left\lfloor \tfrac{r+1}{2} \right\rfloor, \qquad t_0 = t_1 - 1.
$$

Because these consecutive tiles have opposite parity, each row gets exactly one contribution from the even tiles and one from the odd tiles.

**In 2D**, tiles are labeled by row $t_r$ and column $t_c$. Two tiles are disjoint as soon as they are disjoint along one axis. Two different tiles with the same row parity **and** the same column parity always differ by at least 2 in some index, so they are disjoint along that axis. That gives **four classes**, and tiles in one class never overlap. The classes may be labeled as

$$
\text{class}(t_r, t_c) = 2\,(t_r \bmod 2) + (t_c \bmod 2) \in \{0,1,2,3\}
$$

Each class writes its 4 × 4 tiles into its own full-size copy of dx with plain stores, with no conflicts inside a class. An interior pixel is covered by one tile of each of the 4 classes, and the kernel `winconv_3x3_bwd_data_reduce` performs this sum and applies beta.

**Edge cases:** On the border some classes never reach a given pixel, and that plane position was never written, so it holds leftover memory. Instead of zero-filling four planes before every launch, the reduce kernel works out which classes cover the pixel and adds only the contributing planes.

**Cost and when it is used:**

|                        | Extra workspace                       | ResNet-9, batch 128, worst layer          |
| ---------------------- | ------------------------------------- | ----------------------------------------- |
| Backward-filter planes | s · 9 · C · N floats, capped at 64 MB | a few MB                                  |
| Backward-data planes   | 4 × size of dx, capped at 256 MB      | 64 → 128 at 32 × 32: 4 × 33.5 MB = 134 MB |

Above the cap, the code falls back to atomics: at batch 256 that layer's planes would be 268 MB.

**Measured effect and validation:**

The measured effect on the BC-250: ResNet-9 training went from 972 to 1,374 img/s, with gradients matching CPU references. The emulated atomic was 39% of the backward-filter kernel, measured by swapping in a racy +=. The reduce costs under 2% of the kernel it replaces.

-----

Hope this answers your questions. The explicit patch may be found here: https://github.com/mxreyer/pytorch-dlprim-gfx1013/blob/main/patches/dlprimitives/04-winograd-no-atomics.patch (note that it depends on the aforementioned occupancy patch https://github.com/mxreyer/pytorch-dlprim-gfx1013/blob/main/patches/dlprimitives/03-winograd-ksplit-heuristic.patch)

Cheers!
